psi4/driver/driver_findif.py

```diff
@@ -0,0 +1,799 @@
+            for i, index in enumerate(active_indices):
+                for index2 in active_indices[:i]:
+                    for val in data["disps"]["off"]:
+                        append_geoms((index, index2), val)
+
+    disp_geoms.append(ref_geom)
+
+    if data["print_lvl"] > 2:
+        for disp in disp_geoms:
+            disp.print_out()
+
+    if data["print_lvl"] > 1:
+        core.print_out("\n-------------------------------------------------------------\n")
+
+    return disp_geoms
+
+
+def comp_grad_from_energy(mol, E):
+    """Compute the gradient by finite difference of energies.
+
+    Parameters
+    ----------
+    mol : qcdb.molecule or psi4.core.Molecule
+        The molecule to compute the gradient of.
+    E : list of floats
+        A list of energies of the molecule at displaced geometries.
+
+    Returns
+    -------
+    gradient : np.array
+        The gradient in Cartesians, as a matrix with dimensions
+        number-of-atoms by 3. """
+
+    def init_string(data):
+        return ("  Computing gradient from energies.\n"
+                "\tUsing {:d}-point formula.\n"
+                "\tEnergy without displacement: {:15.10f}\n"
+                "\tCheck energies below for precision!\n"
+                "\tForces are for mass-weighted, symmetry-adapted cartesians (in au).\n".format(
+                    data["num_pts"], E[-1]))
+
+    data = _initialize_findif(mol, -1, "1_0", init_string)
+
+    Ndisp = sum(data["Ndisp_pi"]) + 1  # +1 for the reference
+    if len(E) != Ndisp:
+        raise ValidationError("FINDIF: Received {} energies for {} geometries.".format(len(E), Ndisp))
+
+    # Discard the reference energy.
+    E = np.asarray(E[:-1])
+    if data["num_pts"] == 3:
+        g_q = (E[1::2] - E[::2]) / (2.0 * data["disp_size"])
+    elif data["num_pts"] == 5:
+        g_q = (E[0::4] - 8.0 * E[1::4] + 8.0 * E[2::4] - E[3::4]) / (12.0 * data["disp_size"])
+    else:  # This error SHOULD have already been caught, but just in case...
+        raise ValidationError("FINDIF: {} is an invalid number of points.".format(data["num_pts"]))
+    g_q = np.asarray(g_q)
+
+    if data["print_lvl"]:
+        max_disp = (data["num_pts"] - 1) // 2  # The numerator had better be divisible by two.
+        e_per_salc = 2 * max_disp
+        energy_string = ""
+        for i in range(1, max_disp + 1):
+            energy_string = "Energy(-{})        ".format(i) + energy_string + "Energy(+{})        ".format(i)
+        core.print_out("\n\t Coord      " + energy_string + "    Force\n")
+        for salc in range(data["Nsalc"]):
+            print_str = "\t{:5d}" + " {:17.10f}" * (e_per_salc) + " {force:17.10f}" + "\n"
+            energies = E[e_per_salc * salc:e_per_salc * (salc + 1)]
+            print_str = print_str.format(salc, force=g_q[salc], *energies)
+            core.print_out(print_str)
+        core.print_out("\n")
+
+    B = data["salc_list"].matrix()
+    g_cart = np.matmul(g_q, B)
```

Comment:
`matmul` was implemented for the `@` python operator. This is still a bit of an experimental function in NumPy despite it being in many releases. Would use `np.dot` in general.